### PR TITLE
Account based eacl targets

### PR DESCRIPTION
**Pull request by smallhive**
Refs https://github.com/nspcc-dev/neofs-api/issues/278

### CODE REVIEW

File: eacl/target.go [hunk 1]
```diff
@@ -37,23 +37,25 @@ func (t Target) CopyTo(dst *Target) {
...
 }
 
-// BinaryKeys returns list of public keys to identify
+// BinaryKeys returns list of public keys or accounts to identify
 // target subject in a binary format.
 //
-// Each element of the resulting slice is a serialized compressed public key. See [elliptic.MarshalCompressed].
-// Use [neofsecdsa.PublicKey.Decode] to decode it into a type-specific structure.
+// Each element of the resulting slice is a serialized compressed public key or account from the public key.
+// See [elliptic.MarshalCompressed], [keys.PublicKey.GetScriptHash].
```
Comment (cthulhu-rider) [inline eacl/target.go:44]: i see u decided to set accounts as 20B script hashes instead of 25B `user.ID`. Guess @roman-khimov will be incredibly happy about this :]

i agree with the format, but if then i also want to provide proper way to work with `user.ID`s. We already have https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.12/user#ID.SetScriptHash, lets provide also provide getter

**Review by carpawell — COMMENTED** [review]
Why not API repo changes first?

**Comment by cthulhu-rider** [comment]
> Why not API repo changes first?

there will be one comment in docs, no binary protocol changes. But we wanna review the concept in practice. In total, yes, we'll have to merge this and API simultaneously @smallhive

File: eacl/target.go (hunk 1 above)
Comment (smallhive) [reply eacl/target.go:44]: Well, from my point of view, I thought It was the same thing. I have to revalidate these

File: eacl/target.go [hunk 2]
```diff
@@ -37,23 +37,25 @@ func (t Target) CopyTo(dst *Target) {
...
+// in other case.
 //
 // The value returned shares memory with the structure itself, so changing it can lead to data corruption.
 // Make a copy if you need to change it.
 func (t *Target) BinaryKeys() [][]byte {
 	return t.keys
 }
 
-// SetBinaryKeys sets list of binary public keys to identify
+// SetBinaryKeys sets list of binary public keys or accounts to identify
```
[the remark this answers is not in the document]
Comment (smallhive) [reply eacl/target.go]: I've added SetAccounts/Accounts methods. But I think we don't need to deprecate SetKeys methods because they are still valid and required. How I understand the idea, they still required.
Moreover, I am considering adding length checks to be sure it is an real account data. What do you think? @cthulhu-rider

File: user/id.go [hunk 3]
```diff
@@ -23,6 +23,14 @@ type ID struct {
 	w []byte
 }
 
+// WithScriptHash is a constructor for ID, using [util.Uint160].
+func WithScriptHash(scriptHash util.Uint160) ID {
```
Comment (cthulhu-rider) [inline user/id.go:27]:
```suggestion
func NewFromScriptHash(scriptHash util.Uint160) ID {
```
and mark this refs #483, and mark `SetScriptHash` as `Deprecated`

File: user/id.go [hunk 4]
```diff
@@ -67,6 +75,20 @@ func (x *ID) SetScriptHash(scriptHash util.Uint160) {
 	copy(x.w[21:], hash.Checksum(x.w[:21]))
 }
 
+// GetScriptHash gets scripthash from user ID.
+func (x *ID) GetScriptHash() (util.Uint160, error) {
```
Comment (cthulhu-rider) [inline user/id.go:79]: i dont think such signature would be convenient to use. How about
```
// GetScriptHash gets script hash from user ID.
func (x ID) GetScriptHash() util.Uint160 {
	return util.Uint160(x.w[1:21])
}
```
if `ID` constructed correctly - ok, otherwise panic, so be it

File: eacl/target.go [hunk 5]
```diff
@@ -77,6 +99,42 @@ func SetTargetECDSAKeys(t *Target, pubs ...*ecdsa.PublicKey) {
...
+	}
+
+	for i := 0; i < ln; i++ {
+		binKeys = append(binKeys, accs[i].BytesBE())
+	}
+
+	t.SetAccounts(binKeys)
+}
+
+// SetTargetUsers sets user ids in Target.
```
Comment (cthulhu-rider) [inline eacl/target.go:120]:
```suggestion
// SetTargetUsers sets user IDs in Target.
```

File: eacl/target.go [hunk 6]
```diff
@@ -58,6 +60,26 @@ func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a serialized account from the public key.
```
Comment (cthulhu-rider) [inline eacl/target.go:65]: > from the public key

again, not always true - multisigs exist

```suggestion
// Each element of the resulting slice is a 20 byte account script hash.
```

and then add `Instead of a script hash there can also be a 33 byte public key binary.`

File: eacl/validator.go [hunk 7]
```diff
@@ -152,8 +152,14 @@ func targetMatches(unit *ValidationUnit, record *Record) bool {
...
 		if pubs := target.BinaryKeys(); len(pubs) != 0 {
 			for _, key := range pubs {
-				if bytes.Equal(key, unit.key) {
-					return true
+				if len(key) == 33 { // pub key.
+					if bytes.Equal(key, unit.key) {
+						return true
+					}
+				} else { // account (script hash).
+					if bytes.Equal(key, unit.account) {
```
Comment (cthulhu-rider) [inline eacl/validator.go:160]: i'm worried about the case when `unit.account` is unset and `key` is empty (afaik this is not prohibited now). Suggest
```suggestion
					if len(unit.account) > 0 && bytes.Equal(key, unit.account) {
```

File: eacl/target.go [hunk 8]
```diff
@@ -77,6 +99,42 @@ func SetTargetECDSAKeys(t *Target, pubs ...*ecdsa.PublicKey) {
...
+	ln := len(accs)
+
+	if cap(binKeys) >= ln {
+		binKeys = binKeys[:0]
+	} else {
+		binKeys = make([][]byte, 0, ln)
+	}
+
+	for i := 0; i < ln; i++ {
+		binKeys = append(binKeys, accs[i].WalletBytes())
```
Comment (cthulhu-rider) [inline eacl/target.go:132]: again this differs with `SetTargetAccounts`: there u set 20B, here - 25B. Guess u wanted to use `accs[i].GetScriptHash()`

File: eacl/target.go [hunk 9]
```diff
@@ -58,6 +60,26 @@ func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a serialized account from the public key.
+// See [keys.PublicKey.GetScriptHash].
+// Use [address.StringToUint160] to decode it into a type-specific structure.
```
Comment (cthulhu-rider) [inline eacl/target.go:67]: i guess u meant `util.Uint160DecodeBytesBE`

File: eacl/target.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply eacl/target.go]: we provide `SetAccount` / `Accounts` and write in docs that argument _can be_ binary-encoded public key. With this, `*BinaryKeys` become obsolete so we deprecated them

see also https://github.com/nspcc-dev/neofs-sdk-go/pull/592#discussion_r1666734549

File: eacl/validator_test.go [hunk 10]
```diff
@@ -236,6 +265,18 @@ func makeKeys(t *testing.T, n int) [][]byte {
 	return pubs
 }
 
+func accounts(t *testing.T, n int) [][]byte {
+	pubs := make([][]byte, n)
+	for i := range pubs {
+		pubs[i] = make([]byte, 25)
```
Comment (cthulhu-rider) [inline eacl/validator_test.go:271]:
```suggestion
		pubs[i] = make([]byte, 20)
```

File: eacl/validator_test.go [hunk 11]
```diff
@@ -197,6 +197,35 @@ func TestTargetMatches(t *testing.T) {
 	require.False(t, targetMatches(u, r))
 }
 
+func TestTargetMatchesAccount(t *testing.T) {
```
Comment (cthulhu-rider) [inline eacl/validator_test.go:200]: i'd also test mixture of account(s) and public key(s) in one rule/unit

**Comment by smallhive** [comment]
Updated to `user.ID` account meaning with 25 byte length

File: eacl/validator.go [hunk 12]
```diff
@@ -152,8 +152,15 @@ func targetMatches(unit *ValidationUnit, record *Record) bool {
 		// check public key match
 		if pubs := target.BinaryKeys(); len(pubs) != 0 {
 			for _, key := range pubs {
-				if bytes.Equal(key, unit.key) {
-					return true
+				switch len(key) {
+				case 25: // user.ID.
+					if len(unit.account) > 0 && bytes.Equal(key, unit.account) {
```
Comment (roman-khimov) [inline eacl/validator.go:157]: Why `len()` check here?

Comment (smallhive) [reply eacl/validator.go:157]: @cthulhu-rider [worried](https://github.com/nspcc-dev/neofs-sdk-go/pull/592#discussion_r1666737509) about the case when unit.account is unset and key is empty

File: eacl/validator.go [hunk 13]
```diff
@@ -152,8 +152,15 @@ func targetMatches(unit *ValidationUnit, record *Record) bool {
 		// check public key match
 		if pubs := target.BinaryKeys(); len(pubs) != 0 {
 			for _, key := range pubs {
-				if bytes.Equal(key, unit.key) {
-					return true
+				switch len(key) {
+				case 25: // user.ID.
```
Comment (cthulhu-rider) [inline eacl/validator.go:156]:
```suggestion
				case user.IDSize:
```
is now available

File: eacl/validator_test.go [hunk 14]
```diff
@@ -236,6 +312,18 @@ func makeKeys(t *testing.T, n int) [][]byte {
 	return pubs
 }
 
+func accounts(t *testing.T, n int) [][]byte {
+	pubs := make([][]byte, n)
+	for i := range pubs {
+		pubs[i] = make([]byte, 25)
+
+		//nolint:staticcheck
+		_, err := rand.Read(pubs[i][1:])
```
Comment (cthulhu-rider) [inline eacl/validator_test.go:321]: why `[1:]`? maybe just `usertest.IDs(n)`?

File: user/id_test.go [hunk 15]
```diff
@@ -58,6 +59,31 @@ func TestID_ReadFromV2(t *testing.T) {
 	})
 }
 
+func TestID_ScriptHash(t *testing.T) {
```
Comment (cthulhu-rider) [inline user/id_test.go:62]: not needed, `TestNewFromScriptHash` already exists

File: user/id.go [hunk 16]
```diff
@@ -82,6 +82,11 @@ func (x ID) WriteToV2(m *refs.OwnerID) {
 // Deprecated: use [NewFromScriptHash] instead.
 func (x *ID) SetScriptHash(scriptHash util.Uint160) { *x = NewFromScriptHash(scriptHash) }
 
+// GetScriptHash gets scripthash from user ID.
+func (x *ID) GetScriptHash() util.Uint160 {
```
Comment (cthulhu-rider) [inline user/id.go:86]:
```suggestion
func (x ID) GetScriptHash() util.Uint160 {
```

File: eacl/target.go [hunk 17]
```diff
@@ -58,6 +60,27 @@ func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a 25 byte account.
```
Comment (cthulhu-rider) [inline eacl/target.go:65]:
```suggestion
// Each element of the resulting slice is a [user.ID] byte account.
```

File: eacl/target.go [hunk 18]
```diff
@@ -58,6 +60,27 @@ func (t *Target) SetBinaryKeys(keys [][]byte) {
...
+// Use `user := user.ID(slice)` to decode it into a type-specific structure.
+//
+// Instead of an account there can also be a 33 byte public key binary.
+// See [keys.PublicKey].
+//
+// The value returned shares memory with the structure itself, so changing it can lead to data corruption.
+// Make a copy if you need to change it.
+func (t *Target) Accounts() [][]byte {
```
Comment (cthulhu-rider) [inline eacl/target.go:73]: lets deprecate `BinaryKeys` and `SetBinaryKeys` then

File: eacl/target.go (hunk 17 above)
Comment (cthulhu-rider) [inline eacl/target.go:65]: user can do `t.SetAccounts([][]byte{[]byte("Hello, world!")})` making this statement false. I would emphasize that this is the expected length, but to decode it you need to explicitly check it

File: eacl/target.go (hunk 18 above)
Comment (cthulhu-rider) [inline eacl/target.go:73]:
```suggestion
func (t Target) Accounts() [][]byte {
```

File: eacl/target.go [hunk 19]
```diff
@@ -77,6 +100,43 @@ func SetTargetECDSAKeys(t *Target, pubs ...*ecdsa.PublicKey) {
 	t.SetBinaryKeys(binKeys)
 }
 
+// SetTargetAccounts sets accounts in Target.
+func SetTargetAccounts(t *Target, accs ...util.Uint160) {
```
Comment (cthulhu-rider) [inline eacl/target.go:104]: suggest to make like in this [PR](https://github.com/nspcc-dev/neofs-sdk-go/pull/605) - provide constructors `TargetByAccounts`/`TargetByUsers`, do not add this and `SetTargetUsers`. Test functions may also be reflected

File: eacl/types.go [hunk 20]
```diff
@@ -91,3 +93,14 @@ func (u *ValidationUnit) WithEACLTable(table *Table) *ValidationUnit {
 
 	return u
 }
+
+// WithAccount configures ValidationUnit to use as sender's account.
```
Comment (cthulhu-rider) [inline eacl/types.go:97]:
```suggestion
// WithAccount configures ValidationUnit to use as sender's [user.ID] bytes.
```

File: eacl/types.go [hunk 21]
```diff
@@ -91,3 +93,14 @@ func (u *ValidationUnit) WithEACLTable(table *Table) *ValidationUnit {
...
+//
+// Parameter v is a serialized account from sender's public key.
```
Comment (cthulhu-rider) [inline eacl/types.go:99]: redundant statement

File: eacl/validator.go (hunk 12 above)
Comment (cthulhu-rider) [reply eacl/validator.go:157]: > and key is empty

but now u use switch-case so key is definitely not empty

File: eacl/target.go (hunk 17 above)
Comment (smallhive) [reply eacl/target.go]: Updated

File: eacl/target.go (hunk 18 above)
Comment (smallhive) [reply eacl/target.go]: Don't we plan to use pub keys later?

Comment (smallhive) [reply eacl/target.go]: Updated

File: eacl/types.go (hunk 20 above)
Comment (smallhive) [reply eacl/types.go]: Done

File: eacl/types.go (hunk 21 above)
Comment (smallhive) [reply eacl/types.go]: Removed

File: eacl/validator.go (hunk 13 above)
Comment (smallhive) [reply eacl/validator.go]: Done

File: eacl/validator.go (hunk 12 above)
Comment (smallhive) [reply eacl/validator.go]: Removed

File: eacl/validator_test.go (hunk 14 above)
Comment (smallhive) [reply eacl/validator_test.go]: Updated

File: user/id.go (hunk 16 above)
Comment (smallhive) [reply user/id.go]: Done

File: user/id_test.go (hunk 15 above)
Comment (smallhive) [reply user/id_test.go]: Removed

**Comment by carpawell** [comment]
> Why not API repo changes first?

But still.

File: eacl/target.go (hunk 18 above)
Comment (cthulhu-rider) [reply eacl/target.go]: they are already supported by `SetAccounts`

File: eacl/target.go [hunk 22]
```diff
@@ -58,6 +60,27 @@ func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a [user.ID] byte account.
```
Comment (cthulhu-rider) [inline eacl/target.go:65]: sry ive mistaken in previous suggestion
```suggestion
// Each element of the resulting slice is a [user.IDSize] byte account.
```
or just
```suggestion
// Each element of the resulting slice is a [user.ID].
```

File: eacl/target.go (hunk 17 above)
Comment (smallhive) [reply eacl/target.go]: Added

File: eacl/target.go (hunk 18 above)
Comment (smallhive) [reply eacl/target.go]: Done

File: eacl/target.go (hunk 19 above)
Comment (smallhive) [reply eacl/target.go:104]: These methods use existing `t *Target` to extend it. Constructor is not a good fit for this goal

File: eacl/target.go (hunk 22 above)
Comment (smallhive) [reply eacl/target.go]: Updated

File: eacl/target.go [hunk 23]
```diff
@@ -54,10 +57,34 @@ func (t *Target) BinaryKeys() [][]byte {
...
+// The value returned shares memory with the structure itself, so changing it can lead to data corruption.
+// Make a copy if you need to change it.
+func (t Target) Accounts() [][]byte {
+	return t.keys
+}
+
+// SetAccounts sets list of accounts to identify target subject.
+//
+// Each element of the accounts parameter is a slice of bytes of [user.ID].
+func (t *Target) SetAccounts(accounts [][]byte) {
```
Comment (carpawell) [inline eacl/target.go:84]: why not just `user.ID` then? it is even documented like this

File: eacl/target.go [hunk 24]
```diff
@@ -77,6 +104,43 @@ func SetTargetECDSAKeys(t *Target, pubs ...*ecdsa.PublicKey) {
 	t.SetBinaryKeys(binKeys)
 }
 
+// SetTargetAccounts sets accounts in Target.
+func SetTargetAccounts(t *Target, accs ...util.Uint160) {
```
Comment (carpawell) [inline eacl/target.go:108]: is this method expected to be used? what cases?

File: eacl/types.go [hunk 25]
```diff
@@ -83,11 +85,20 @@ func (u *ValidationUnit) WithSenderKey(v []byte) *ValidationUnit {
 	return u
 }
 
-// WithBearerToken configures ValidationUnit to use v as request's bearer token.
+// WithEACLTable configures ValidationUnit to use v as request's bearer token.
```
Comment (carpawell) [inline eacl/types.go:88]: non-functional change that does not relate commit, do it in a separate commit?

File: user/id.go [hunk 26]
```diff
@@ -82,6 +82,11 @@ func (x ID) WriteToV2(m *refs.OwnerID) {
 // Deprecated: use [NewFromScriptHash] instead.
 func (x *ID) SetScriptHash(scriptHash util.Uint160) { *x = NewFromScriptHash(scriptHash) }
 
+// GetScriptHash gets scripthash from user ID.
+func (x ID) GetScriptHash() util.Uint160 {
```
Comment (carpawell) [inline user/id.go:86]: is used only in tests in this PR, also an additional commit?

File: eacl/target.go (hunk 23 above)
Comment (smallhive) [reply eacl/target.go:84]: `SetBinaryKeys` is deprecated now and changing `SetAccounts` signature eliminates possibility to pass public keys

File: eacl/target.go (hunk 24 above)
Comment (smallhive) [reply eacl/target.go:108]: We suggest SDK and from my point of view it would be useful for users, for case when you don't need to convert util.Uint160 to user.ID

File: user/id.go (hunk 26 above)
Comment (smallhive) [reply user/id.go:86]: Moved to additional commit

File: eacl/types.go (hunk 25 above)
Comment (smallhive) [reply eacl/types.go:88]: Moved to an additional commit

File: eacl/target.go (hunk 23 above)
Comment (carpawell) [reply eacl/target.go:84]: do not see problems or do not understand. if we do not want users to use `SetBinaryKeys` why we want them to be able to pass keys via `SetAccounts` (a func that is even documented to be used with user.ID slices only)?

File: eacl/target.go (hunk 24 above)
Comment (carpawell) [reply eacl/target.go:108]: @cthulhu-rider, what do you think about it? (you usually try to drop code whose usage is not clear). i just not sure where it can be used

Comment (cthulhu-rider) [reply eacl/target.go:108]: in #605 i deprecate them and provide
```go
func TargetByPublicKeys(ks [][]byte) Target
```
(there are no users introduced yet). With this, i'd provide additional
```
func TargetByUsers(us []user.ID) Target
func TargetByAccounts(hs []util.Uint160) Target
```

File: eacl/target.go (hunk 23 above)
Comment (cthulhu-rider) [reply eacl/target.go:84]: i bet the reason is that we cannot keep getter and setter symmetry while having `Set([]user.ID)`. Field is `[][]byte` anyway, so it's pretty normal to accept/return value of this type. At the same time, we may provide functional helpers for those who desires to set `[]user.ID` (https://github.com/nspcc-dev/neofs-sdk-go/pull/592#discussion_r1696720643)

Comment (carpawell) [reply eacl/target.go:84]: > Field is [][]byte anyway

the field should be kept as long as we have such eALCs set in public networks. but new eACLs should be filled with `SetAccounts` and no keys should be accepted. at least that is how i see it

Comment (cthulhu-rider) [reply eacl/target.go:84]: > new eACLs should be filled with SetAccounts and no keys should be accepted

it's questionable which ones are _new_. For example, if i need to add one rule to my existing table w/o touching previously set records - i dont expect the system to deny such table. It worked with N records, i expect it to work with N+1 ones. Yeah, i _can_ refill they whole table, but this is inconvenient for me being a client

> the field should be kept as long as we have such eALCs set in public networks

weve been through this before - its impossible to track this in general. We need to clearly define the moment after which support will be cut. This should be either a system timestamp or next major protocol version

Comment (carpawell) [reply eacl/target.go:84]: why we deprecate if we gonna support it? to me, that should work this way: 0.43.0 release (node and ir) supports both ways, 0.44.0 supports only accounts. "supports" means accepting new tables, old ones should always work until we understand there are no such tables in the contract

anyway, what is your understanding of public keys deprecation? how long are we gonna support it? till 1.0.0 SDK release? what definition will it have? `SetAccounts(accounts [][]byte)` too?

Comment (roman-khimov) [reply eacl/target.go:84]: Keys will be supported for as long as needed.

Comment (roman-khimov) [reply eacl/target.go:84]: As for the method interface --- I'd really love to see more specific type here. It can be done symmetrically, the only problem is keys in existing eacls. Keys can be converted into standard accounts, but that can be misleading. We can treat them as separate entities though. We already have `Role` concept in the `Target` that is independent of `Keys`. So we can also have `Accounts` that are independent of them, you get keys (size-checked!) in `BinaryKeys()` and you get accounts from `Accounts()`, they can technically be mixed in the same eacl rule.

File: user/id.go (hunk 26 above)
Comment (roman-khimov) [inline user/id.go:86]: Can just be `ScriptHash()`, no need to `Get`.

File: eacl/target.go [hunk 27]
```diff
@@ -45,6 +47,7 @@ func (t Target) CopyTo(dst *Target) {
 //
 // The value returned shares memory with the structure itself, so changing it can lead to data corruption.
 // Make a copy if you need to change it.
+// Deprecated: use [Target.Accounts] instead.
 func (t *Target) BinaryKeys() [][]byte {
 	return t.keys
```
Comment (cthulhu-rider) [inline eacl/target.go:52]: implementation of this method must be changed for backward compatibility. Some `t.keys` items can be 25B now, so they must be excluded

**Review by cthulhu-rider — CHANGES_REQUESTED** [review]

File: eacl/target.go (hunk 23 above)
Comment (cthulhu-rider) [reply eacl/target.go:84]: all cases can be supported by functions around the most generic methods accepting/returning `[][]byte`, so currently proposed approach is the most preferable to me cuz it is the closest to the protocol

File: eacl/target.go [hunk 28]
```diff
@@ -54,10 +75,41 @@ func (t *Target) BinaryKeys() [][]byte {
...
 func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a [user.IDSize] byte account.
+// Use `user := user.ID(slice)` to decode it into a type-specific structure.
+//
+// The value returned shares memory with the structure itself, so changing it can lead to data corruption.
```
Comment (cthulhu-rider) [inline eacl/target.go:88]: not critical, but it does not share memory, copying would be an overhead

**Review by cthulhu-rider — APPROVED** [review]

File: eacl/target.go [hunk 29]
```diff
@@ -45,7 +47,26 @@ func (t Target) CopyTo(dst *Target) {
...
+	return r
+}
+
+// RawSubjects returns list of public keys or [user.ID] to identify target subject in a binary format.
+//
+// If element length is 33, it is a serialized compressed public key. See [elliptic.MarshalCompressed], [keys.PublicKey.GetScriptHash].
+// If element length is 25, it is a [user.ID]. Use `id :=user.ID(element)`.
+//
+// Using this method is your responsibility.
+func (t *Target) RawSubjects() [][]byte {
```
Comment (cthulhu-rider) [inline eacl/target.go:69]:
```suggestion
func (t Target) RawSubjects() [][]byte {
```

Comment (smallhive) [reply eacl/target.go]: Done

File: eacl/target.go (hunk 28 above)
Comment (smallhive) [reply eacl/target.go]: Removed

File: eacl/target.go [hunk 30]
```diff
@@ -45,7 +47,26 @@ func (t Target) CopyTo(dst *Target) {
 //
 // The value returned shares memory with the structure itself, so changing it can lead to data corruption.
 // Make a copy if you need to change it.
+// Deprecated: use [Target.Accounts] instead.
```
Comment (carpawell) [inline eacl/target.go:50]: sounds a little strange to me: if i am using binary keys, using `Accounts` does not solve my "deprication-message-from-linter" problem; can be like "using raw binary keys is not encouraged and will be dropped eventually, use Accoun-based targets instead". however, it is a minor problem

File: eacl/target.go [hunk 31]
```diff
@@ -45,7 +47,26 @@ func (t Target) CopyTo(dst *Target) {
...
+		}
+	}
+
+	return r
+}
+
+// RawSubjects returns list of public keys or [user.ID] to identify target subject in a binary format.
+//
+// If element length is 33, it is a serialized compressed public key. See [elliptic.MarshalCompressed], [keys.PublicKey.GetScriptHash].
+// If element length is 25, it is a [user.ID]. Use `id :=user.ID(element)`.
```
Comment (carpawell) [inline eacl/target.go:66]:
```suggestion
// If element length is 25, it is a [user.ID]. Use `id := user.ID(element)`.
```
to be symmetrical

File: eacl/target.go [hunk 32]
```diff
@@ -54,10 +75,38 @@ func (t *Target) BinaryKeys() [][]byte {
...
 // Each element of the keys parameter is a slice of bytes is a serialized compressed public key.
 // See [elliptic.MarshalCompressed].
+// Deprecated: use [Target.SetAccounts] instead.
 func (t *Target) SetBinaryKeys(keys [][]byte) {
 	t.keys = keys
 }
 
+// Accounts returns list of accounts to identify target subject in a binary format.
+//
+// Each element of the resulting slice is a [user.IDSize] byte account.
```
Comment (carpawell) [inline eacl/target.go:85]: two outdated lines?

File: eacl/target.go [hunk 33]
```diff
@@ -54,10 +75,38 @@ func (t *Target) BinaryKeys() [][]byte {
...
+			r = append(r, user.ID(key))
+		}
+	}
+
+	return r
+}
+
+// SetAccounts sets list of accounts to identify target subject.
+//
+// Each element of the accounts parameter is a slice of bytes of [user.ID].
```
Comment (carpawell) [inline eacl/target.go:101]: outdated line?

File: eacl/types.go [hunk 34]
```diff
@@ -92,3 +94,12 @@ func (u *ValidationUnit) WithEACLTable(table *Table) *ValidationUnit {
 
 	return u
 }
+
+// WithAccount configures ValidationUnit to use as sender's [user.ID] bytes.
+func (u *ValidationUnit) WithAccount(v []byte) *ValidationUnit {
```
Comment (carpawell) [inline eacl/types.go:99]: shouldnt it be `user.ID` too? why it differs from `Target`?

File: eacl/target.go (hunk 31 above)
Comment (smallhive) [reply eacl/target.go]: Done

File: eacl/target.go (hunk 33 above)
Comment (smallhive) [reply eacl/target.go]: Removed

File: eacl/types.go (hunk 34 above)
Comment (smallhive) [reply eacl/types.go]: yes, it should. Updated

File: eacl/target.go (hunk 32 above)
Comment (smallhive) [reply eacl/target.go]: Updated and removed one of them

**Review by carpawell — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]
